# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: ipc/glue/UtilityProcessManager.cpp
# Commit: af562cec3d70
# Full Hash: af562cec3d706ee3ed605ffeade4d3ba982d2545
# Author: Ray Kraesig <rkraesig@mozilla.com>
# Date: 2023-10-27 04:08:13
# Regressor Bug: 1837079
# File Overlap Count: 2
# Description:
#   Bug 1837079 - [5/10] Create WinFileDialog IPDL protocol and implementation  r=gstoll,handyman,ipc-reviewers,nika,win-reviewers,mhowell
#   
#   Create and implement a new top-level IPC protocol, `PWinFileDialog`,
#   using the primitives from bug 1833450 and the new sandboxing type from
#   the previous commit.
# ==============================================================================

diff -r b5d52e5f75f2 -r af562cec3d70 ipc/glue/UtilityProcessManager.cpp
--- a/ipc/glue/UtilityProcessManager.cpp	Thu Oct 26 18:21:28 2023 +0000
+++ b/ipc/glue/UtilityProcessManager.cpp	Thu Oct 26 18:21:28 2023 +0000
@@ -24,6 +24,7 @@
 
 #ifdef XP_WIN
 #  include "mozilla/dom/WindowsUtilsParent.h"
+#  include "mozilla/widget/filedialog/WinFileDialogParent.h"
 #endif
 
 #include "mozilla/GeckoArgs.h"
@@ -449,6 +450,42 @@
 
 void UtilityProcessManager::ReleaseWindowsUtils() { mWindowsUtils = nullptr; }
 
+RefPtr<UtilityProcessManager::WinFileDialogPromise>
+UtilityProcessManager::CreateWinFileDialogAsync() {
+  using Promise = WinFileDialogPromise;
+  TimeStamp startTime = TimeStamp::Now();
+  auto wfdp = MakeRefPtr<widget::filedialog::WinFileDialogParent>();
+
+  return StartUtility(wfdp, SandboxingKind::WINDOWS_FILE_DIALOG)
+      ->Then(
+          GetMainThreadSerialEventTarget(), __PRETTY_FUNCTION__,
+          [wfdp, startTime]() {
+            LOGD("CreateWinFileDialogAsync() resolve: wfdp = [%p]", wfdp.get());
+            if (!wfdp->CanSend()) {
+              MOZ_ASSERT(false, "WinFileDialogParent can't send");
+              return Promise::CreateAndReject(NS_ERROR_FAILURE,
+                                              __PRETTY_FUNCTION__);
+            }
+            PROFILER_MARKER_TEXT(
+                "UtilityProcessManager::CreateWinFileDialogAsync", OTHER,
+                MarkerOptions(MarkerTiming::IntervalUntilNowFrom(startTime)),
+                "Resolve"_ns);
+            return Promise::CreateAndResolve(wfdp, __PRETTY_FUNCTION__);
+          },
+          [self = RefPtr(this), startTime](nsresult error) {
+            LOGD("CreateWinFileDialogAsync() reject");
+            if (!self->IsShutdown()) {
+              MOZ_ASSERT_UNREACHABLE("failure when starting file-dialog actor");
+            }
+            PROFILER_MARKER_TEXT(
+                "UtilityProcessManager::CreateWinFileDialogAsync", OTHER,
+                MarkerOptions(MarkerTiming::IntervalUntilNowFrom(startTime)),
+                "Reject"_ns);
+
+            return Promise::CreateAndReject(error, __PRETTY_FUNCTION__);
+          });
+}
+
 #endif  // XP_WIN
 
 bool UtilityProcessManager::IsProcessLaunching(SandboxingKind aSandbox) {